clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -314,69 +370,118 @@ public void poll(long timeout) throws IOException {
+                        //(possibly because no memory). it may be the case that the underlying socket will
+                        //not come up in the next poll() and so we need to remember this channel for the
+                        //next poll call otherwise data may be stuck in said buffers forever.
+                        keysWithBytesBuffered.add(key);
+                    }
 
-                /* if channel is ready read from any connections that have readable data */
-                if (channel.ready() && key.isReadable() && !hasStagedReceive(channel)) {
-                    NetworkReceive networkReceive;
-                    while ((networkReceive = channel.read()) != null)
-                        addToStagedReceives(channel, networkReceive);
-                }
+                    /* if channel is ready write to any sockets that have space in their buffer and for which we have data */
+                    if (channel.ready() && key.isWritable()) {
+                        madeProgress = true;
+                        Send send = channel.write();
+                        if (send != null) {
+                            this.completedSends.add(send);
+                            this.sensors.recordBytesSent(channel.id(), send.size());
+                        }
+                    }
 
-                /* if channel is ready write to any sockets that have space in their buffer and for which we have data */
-                if (channel.ready() && key.isWritable()) {
-                    Send send = channel.write();
-                    if (send != null) {
-                        this.completedSends.add(send);
-                        this.sensors.recordBytesSent(channel.id(), send.size());
+                    /* cancel any defunct sockets */
+                    if (!key.isValid()) {
+                        madeProgress = true;
+                        close(channel, true);
                     }
-                }
 
-                /* cancel any defunct sockets */
-                if (!key.isValid())
+                } catch (Exception e) {
+                    String desc = channel.socketDescription();
+                    if (e instanceof IOException)
+                        log.debug("Connection with {} disconnected", desc, e);
+                    else
+                        log.warn("Unexpected error from {}; closing connection", desc, e);
                     close(channel, true);
-
-            } catch (Exception e) {
-                String desc = channel.socketDescription();
-                if (e instanceof IOException)
-                    log.debug("Connection with {} disconnected", desc, e);
-                else
-                    log.warn("Unexpected error from {}; closing connection", desc, e);
-                close(channel, true);
+                }
             }
+            return madeProgress;
+        } finally {
+            if (keysHandled.size() == selectionKeys.size()) {
```

Comment:
the sizes will differ only if some uncaught exception terminates the loop early (so never, unless bug?). the code path for different sizes is there to try and match what the previous code would produce under those conditions (which are, again, a bug).

under "normal" operating conditions the sizes should always be the same, which i think makes the clear() calls a faster implementation (N * arrayList.add() + 2*clear() < N * set.Iterator.remove()).

what do you mean by "ready for selection table"? looking at the code for openjdk 8 selectors use a normal set, wrapped to disallow external add() calls. all selectors do is call add() on the set of keys, there's no special "hook" to react to removes/clears